test/app-luatest/console_debugger_session_test.lua

```diff
@@ -0,0 +1,134 @@
+local popen = require('popen')
+local tnt = require('tarantool')
+local g = t.group()
+
+local function normalize_path(s)
+    return s:gsub("^@", ""):gsub("[^/]+$", "")
+end
+
+local function unescape(s)
+    return s:gsub('[\27\155][][()#;?%d]*[A-PRZcf-ntqry=><~]', '')
+end
+
+local function trim(s)
+    return s:gsub('%s+$', '')
+end
+
+local function tarantool_path(arg)
+    local index = -2
+    -- arg[-1] is guaranteed to be non-null
+    while arg[index] do index = index - 1 end
+    return arg[index + 1]
+end
+
+local TARANTOOL_PATH = tarantool_path(arg)
+local path_to_script = normalize_path(debug.getinfo(1, 'S').source)
+local debug_target_script = path_to_script .. 'debug-target.lua'
+
+local DEBUGGER = 'luadebug.lua'
+local dbg_header = DEBUGGER .. ": Loaded for " .. tnt.version
+local dbg_prompt = DEBUGGER .. '>'
+
+local cmd_aliases = {
+    ['c'] = 'c|cont|continue',
+    ['d'] = 'd|down',
+    ['e'] = 'e|eval',
+    ['f'] = 'f|finish|step_out',
+    ['h'] = 'h|help|?',
+    ['l'] = 'l|locals',
+    ['n'] = 'n|next|step_over',
+    ['p'] = 'p|print',
+    ['q'] = 'q|quit',
+    ['s'] = 's|st|step|step_into',
+    ['t'] = 't|trace|bt',
+    ['u'] = 'u|up',
+    ['w'] = 'w|where',
+}
+
+local MAX_ALIASES_COUNT = 4
+
+local function get_cmd_alias(cmd, number)
+    if not cmd_aliases[cmd] then
+        return cmd
+    end
+    local cmds = {}
+    for v, _ in cmd_aliases[cmd]:gmatch('[^|]+') do
+        table.insert(cmds, v)
+    end
+    return cmds[number % #cmds + 1]
+end
+
+--[[
+    Should handle equally well both:
+    'n' and 'p obj'
+]]
+local function get_key_arg_pair(cmd)
+    local gen = cmd:gmatch('[^%s]+')
+    return gen(), gen()
+end
+
+local sequence = {
+    { ['\t'] = dbg_header }, -- \t is a special value for start
+    { ['n'] = dbg_prompt },
+    { ['s'] = dbg_prompt },
+    { ['n'] = dbg_prompt },
+    { ['n'] = dbg_prompt },
+    { ['p obj'] = 'obj => {"tzoffset" = "+0300", "hour" = 3}' },
+    { ['n'] = dbg_prompt },
+    { ['p ymd'] = 'ymd => false' },
+    { ['w'] = 'local hms = false' },
+    { ['h'] = dbg_prompt },
+    { ['t'] = 'debug-target.lua:5 in chunk at' },
+    { ['u'] = 'Already at the top of the stack.' },
+    { ['d'] = 'debug-target.lua:5 in chunk at' },
+    { ['u'] = 'Inspecting frame: builtin/datetime.lua' },
+    { ['l'] = '{"tzoffset" = "+0300", "hour" = 3}' },
+    { ['f'] = dbg_prompt },
+    { ['n'] = dbg_prompt },
+    { ['p T'] = 'T => 1970-01-01T03:00:00+0300' },
+    { ['n'] = dbg_prompt },
+    { [''] = dbg_prompt },
+    { [''] = dbg_prompt },
+    { [''] = dbg_prompt },
+    { ['p S'] = 'S => "1970-01-01T0300+0300"' },
+    { ['c'] = '' },
+}
+
+g.test_interactive_debugger_session = function()
+    local cmd = { TARANTOOL_PATH, debug_target_script }
+    --[[
+        repeat multiple times to check all command aliases
+    ]]
+    for i=1,MAX_ALIASES_COUNT do
+        local fh = popen.new(cmd, {
+            stdout = popen.opts.PIPE,
+            stderr = popen.opts.DEVNULL,
```

Comment:
Fixed:
```diff
diff --git a/test/app-luatest/console_debugger_session_test.lua b/test/app-luatest/console_debugger_session_test.lua
index c0522d5a0..f747e6f4e 100644
--- a/test/app-luatest/console_debugger_session_test.lua
+++ b/test/app-luatest/console_debugger_session_test.lua
@@ -104,7 +104,7 @@ g.test_interactive_debugger_session = function()
     for i=1,MAX_ALIASES_COUNT do
         local fh = popen.new(cmd, {
             stdout = popen.opts.PIPE,
-            stderr = popen.opts.DEVNULL,
+            stderr = popen.opts.PIPE,
             stdin = popen.opts.PIPE,
         })
         t.assert_is_not(fh, nil)
```

Reply:
It is not used

Reply:
Then it should be empty, right?

Reply:
Yup, that's why it was ignored to /dev/null before recently. But I agreed that it might be not a good idea to ignore entirely whole stderr output, thus mixing it to all from stdout (so eventually, when something would be passed via error elsewhere would be intercepted by test).

But at the moment `luadebug.lua` outputs only to the standard output file, even in the case of error.

Reply:
It looks reasonable for me to check that `stderr` is empty.

Reply:
Well, it took longer than expected, because I tried to adopt approach used in `interactive_tarantool.lua` helper, which is not very easy from the first approach (and from the second also is not easy). After realization that despite neing redirected to the same `popen.opts.PIPE` we do not read _both_ simultaneously (only either stdou, or stderr with `{stderr = true}` option defined), I've came to this simple workaround:

```diff
diff --git a/test/app-luatest/console_debugger_session_test.lua b/test/app-luatest/console_debugger_session_test.lua
index a9697ef71..077545ecd 100644
--- a/test/app-luatest/console_debugger_session_test.lua
+++ b/test/app-luatest/console_debugger_session_test.lua
@@ -123,8 +123,11 @@ g.test_interactive_debugger_session = function()
 
             local result
             local clean_cmd = trim(cmd)
+            -- there should be empty stderr - check it before stdout
+            local errout = fh:read({ timeout = 0.05, stderr = true})
+            t.assert(errout == nil or trim(errout) == '')
             repeat
-                result = trim(unescape(fh:read({ timeout = 1.0 })))
+                result = trim(unescape(fh:read({ timeout = 0.5 })))
             until result ~= '' and result ~= clean_cmd
             if expected ~= '' then
                 t.assert_str_contains(result, expected, false)
```

This could not be a generic solution if we expect variable length of stream (which migt overflow eventually) coming from `stderr`, but in this case we need only check not empty content once, and then bail out if there is some.

The only drawback of current approach - is that slowdown introduced by used `timeout = 0.05`: in the good context, when there is no error message generated, we always wait til the end of timeout period, this always wait 0.05 secs * number_of_steps. (On the other hand, reading from `stdin` using `timeout = 0.5` is almost immediate as there is always content generated).